### PR TITLE
authenticate via facebook

### PR DESCRIPTION
**Pull request by nguyenhad**

### CODE REVIEW

File: app/views/layouts/_header.html.erb [hunk 1]
```diff
@@ -31,6 +31,10 @@
           <li>
             <%= link_to t("views.button.login"), new_user_session_path  %>
           </li>
+          <li>
+            <%= link_to "Sign in with Facebook",
```
Comment (NghiaNTT) [inline app/views/layouts/_header.html.erb]: I18n

**Comment by NghiaNTT** [comment]
Ok.

File: app/controllers/omniauth_callbacks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def facebook
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+
+    if @user.persisted?
+      sign_in_and_redirect @user, :event => :authentication
```
Comment (dieunb) [inline app/controllers/omniauth_callbacks_controller.rb]: convention

File: config/routes.rb [hunk 3]
```diff
@@ -6,7 +6,7 @@
   get "about"   => "static_pages#about"
   get "contact" => "static_pages#contact"
 
-  devise_for :users
+  devise_for :users, controllers: {:omniauth_callbacks => "omniauth_callbacks"}
```
Comment (dieunb) [inline config/routes.rb]: ` {omniauth_callbacks: "omniauth_callbacks"}`

File: app/controllers/omniauth_callbacks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def facebook
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+    if @user.persisted?
+      sign_in_and_redirect @user, :event => :authentication
```
Comment (dieunb) [inline app/controllers/omniauth_callbacks_controller.rb]: `event: :authentication`

File: app/models/user.rb [hunk 5]
```diff
@@ -37,4 +38,23 @@ def unfollow other_user
...
+  class << self
+    def from_omniauth auth
+      where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+        user.email = auth.info.email
+        user.password = Devise.friendly_token[0,20]
+        user.name = auth.info.name
+      end
+    end
+
+    def new_with_session(params, session)
```
Comment (dieunb) [inline app/models/user.rb]: remove `()`

**Pull request merged**